src/OpenTelemetry/ReadOnlyFilteredTagCollection.cs

```diff
@@ -0,0 +1,125 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+using System.Diagnostics;
+#if EXPOSE_EXPERIMENTAL_FEATURES && NET8_0_OR_GREATER
+using System.Diagnostics.CodeAnalysis;
+using OpenTelemetry.Internal;
+#endif
+
+namespace OpenTelemetry;
+
+#if EXPOSE_EXPERIMENTAL_FEATURES
+/// <summary>
+/// A read-only collection of tag key/value pairs which returns a filtered
+/// subset of tags when enumerated.
+/// </summary>
+// Note: Does not implement IReadOnlyCollection<> or IEnumerable<> to
+// prevent accidental boxing.
+#if NET8_0_OR_GREATER
+[Experimental(DiagnosticDefinitions.ExemplarExperimentalApi, UrlFormat = DiagnosticDefinitions.ExperimentalApiUrlFormat)]
+#endif
+public
+#else
+internal
+#endif
+    readonly struct ReadOnlyFilteredTagCollection
+{
+    private readonly HashSet<string>? keyFilter;
+    private readonly KeyValuePair<string, object?>[] tags;
+    private readonly int count;
+
+    internal ReadOnlyFilteredTagCollection(
+        HashSet<string>? keyFilter,
+        KeyValuePair<string, object?>[] tags,
+        int count)
+    {
+        Debug.Assert(tags != null, "tags was null");
+        Debug.Assert(count <= tags!.Length, "count was invalid");
+
+        this.keyFilter = keyFilter;
+        this.tags = tags;
+        this.count = count;
+    }
+
+    /// <summary>
+    /// Gets the maximum number of tags in the collection.
+    /// </summary>
+    /// <remarks>
+    /// Note: Enumerating the collection may return fewer results depending on
+    /// the filter.
+    /// </remarks>
+    public int MaximumCount => this.count;
+
+    /// <summary>
+    /// Returns an enumerator that iterates through the tags.
+    /// </summary>
+    /// <returns><see cref="Enumerator"/>.</returns>
+    public Enumerator GetEnumerator() => new(this);
+
+    internal IReadOnlyList<KeyValuePair<string, object?>> ToReadOnlyList()
+    {
+        var list = new List<KeyValuePair<string, object?>>(this.MaximumCount);
+
+        foreach (var item in this)
+        {
+            list.Add(item);
+        }
+
+        return list;
+    }
+
+    /// <summary>
+    /// Enumerates the elements of a <see cref="ReadOnlyTagCollection"/>.
+    /// </summary>
+    // Note: Does not implement IEnumerator<> to prevent accidental boxing.
+    public struct Enumerator
+    {
+        private readonly ReadOnlyFilteredTagCollection source;
+        private int index;
+
+        internal Enumerator(ReadOnlyFilteredTagCollection source)
+        {
+            this.source = source;
```

Comment:
would it make sense to special case the common case where no view is used, and hence `HashSet<string>? keyFilter` is null, and this will be empty iterator?

Reply:
I don't think it would really do much for perf. If there is no view `ReadOnlyFilteredTagCollection` will essentially be constructed as: `new ReadOnlyFilteredTagCollection(keyFilter: null, tags: Array.Empty<KeyValuePair<string, object?>>(), count: 0)`.

`foreach` over that will do:

```csharp
var enumerator = readOnlyFilteredTagCollection.GetEnumerator();
while (enumerator.MoveNext()) // Will return false
{
    ...
}
```

We could code exporters to do this:

```csharp
var tags = exemplar.FilteredTags;
if (tags.MaximumCount > 0)
{
    foreach (var tag in tags)
    {
         ...do work...
    }
}
```

That would skip even the enumerator part.

I did make this tweak which should help a little bit: https://github.com/open-telemetry/opentelemetry-dotnet/pull/5364/commits/c2c7f0a42ee61d809a7a44f81b10397df205cf24